runtime/interpreter/interpreter_expression.go

```diff
@@ -1076,14 +1077,12 @@ func (interpreter *Interpreter) VisitCastingExpression(expression *ast.CastingEx
 			}
 
 			// The failable cast may upcast to an optional type, e.g. `1 as? Int?`, so box
-			value = interpreter.BoxOptional(locationRange, value, expectedType)
+			value = interpreter.ConvertAndBox(locationRange, value, valueSemaType, expectedType)
```

Comment:
I guess this change is unrelated to entitlements? (vaguely remember we discussed this before?)

Reply:
It is related yes; the reason for this change is that we want casting an entitled reference to always "convert" the set of entitlements on the runtime value to the target type of the cast; i.e. it shouldn't be possible to "hide" an entitlement at runtime. It's the difference between this code having `downDownRef` be `nil` and non-`nil`:

```
let ref = &1 as auth(X) &Int
let anyStruct = ref as AnyStruct
let downRef = (anyStruct as? &Int)!
let downDownRef = downRef as? auth(X) &Int
```